### PR TITLE
Implement streaming iterators

### PR DESCRIPTION
**Pull request by msm-code**
Two new database commands:

```
select into iterator "query";
iterator "name" pop 30;
```

### CODE REVIEW

File: DatabaseSnapshot.cpp [hunk 1]
```diff
@@ -8,11 +8,41 @@
...
+}
+
+DatabaseName DatabaseName::derive_temporary() const {
+    std::string id = random_hex_string(8);
+    std::string fname = "temp." + id + ".ursa";
+    return DatabaseName(db_base, "temp", id, fname);
+}
+
+// TODO: this method should not exist, convert all legacy name occurences
+// to DatbaseName.
```
Comment (chivay) [inline DatabaseSnapshot.cpp:25]: typos

File: Database.cpp [hunk 2]
```diff
@@ -184,11 +215,17 @@ void Database::commit_task(uint64_t task_id) {
...
             }
             ds->toggle_taint(change.parameter);
-            ds->on_disk_metadata_update();
+            ds->save();
+        } else if (change.type == DbChangeType::NewIterator) {
+            DatabaseName itname = DatabaseName::parse(db_base, change.obj_name);
+            load_iterator(itname);
+        } else if (change.type == DbChangeType::UpdateIterator) {
+            DatabaseName itname = DatabaseName::parse(db_base, change.obj_name);
+            uint64_t offset = std::atoi(change.parameter.c_str());
```
Comment (chivay) [inline Database.cpp:224]: :concerned:

File: DatabaseSnapshot.cpp [hunk 3]
```diff
@@ -27,6 +57,31 @@ const OnDiskDataset *DatabaseSnapshot::find_dataset(const std::string &name) con
...
+    int count,
+    std::vector<std::string> *out
+) const {
+    auto it = iterators.find(iterator_id);
+    if (it == iterators.end()) {
+        throw std::runtime_error("tried to read non-existent iterator");
+    }
+
+    // TODO maybe should busy-wait on failure?
+    db_handle.request_iterator_lock(iterator_id);
```
Comment (icedevml) [inline DatabaseSnapshot.cpp:73]: Doesn't it mean that right now concurrent iterator access will fail with an error and the user has to be aware of that?

Comment (icedevml) [reply DatabaseSnapshot.cpp:73]: as far as I see this is not handled right now in https://github.com/CERT-Polska/mquery/pull/59/files

File: Database.cpp (hunk 2 above)
Comment (msm-code) [reply Database.cpp]: you think that was concerning? think again

**Pull request merged**